twinify/napsu_mq/napsu_mq.py

```diff
@@ -0,0 +1,186 @@
+from twinify.napsu_mq.markov_network_jax import MarkovNetworkJax
+from twinify.napsu_mq.marginal_query import FullMarginalQuerySet
+from twinify.napsu_mq.dataframe_data import DataFrameData
+import twinify.napsu_mq.privacy_accounting as privacy_accounting
+import twinify.napsu_mq.maximum_entropy_inference as mei
+
+from d3p.random import PRNGState
+
+
+class NapsuMQModel(InferenceModel):
+
+    def __init__(self) -> None:
+        super().__init__()
+
+    def fit(self, data: pd.DataFrame, rng: PRNGState, epsilon: float, delta: float, **kwargs) -> 'NapsuMQResult':
+        column_feature_set = kwargs['column_feature_set'] if 'column_feature_set' in kwargs else []
+        use_laplace_approximation = kwargs[
+            'use_laplace_approximation'] if 'use_laplace_approximation' in kwargs else False
+
+        dataframe = DataFrameData(data)
+        category_mapping = DataFrameData.get_category_mapping(data)
+        n, d = dataframe.int_array.shape
+        domain_key_list = list(dataframe.values_by_col.keys())
+        domain_value_count_list = [len(dataframe.values_by_col[key]) for key in domain_key_list]
+        domain = Domain(domain_key_list, domain_value_count_list)
+        query_sets = MST_selection(Dataset(dataframe.int_df, domain), epsilon, delta,
+                                   cliques_to_include=column_feature_set)
+        queries = FullMarginalQuerySet(query_sets, dataframe.values_by_col)
+        queries = queries.get_canonical_queries()
+        mnjax = MarkovNetworkJax(dataframe.values_by_col, queries)
+        suff_stat = np.sum(queries.flatten()(dataframe.int_array), axis=0)
+        sensitivity = np.sqrt(2 * len(query_sets))
+        sigma_DP = privacy_accounting.sigma(epsilon, delta, sensitivity)
+        dp_suff_stat = jnp.asarray(np.random.normal(loc=suff_stat, scale=sigma_DP))
+
+        if use_laplace_approximation is True:
+            laplace_approx, success = mei.run_numpyro_laplace_approximation(rng, dp_suff_stat, n, sigma_DP, mnjax)
+            mcmc, backtransform = mei.run_numpyro_mcmc_normalised(
+                rng, dp_suff_stat, n, sigma_DP, mnjax, laplace_approx, num_samples=2000, num_warmup=800, num_chains=4
+            )
+            inf_data = az.from_numpyro(mcmc, log_likelihood=False)
+            posterior_values = inf_data.posterior.stack(draws=("chain", "draw"))
+            posterior_values = backtransform(posterior_values.norm_lambdas.values.transpose())
+
+        else:
+            mcmc = mei.run_numpyro_mcmc(
+                rng, dp_suff_stat, n, sigma_DP, mnjax, num_samples=2000, num_warmup=800, num_chains=4
+            )
+            inf_data = az.from_numpyro(mcmc, log_likelihood=False)
+            posterior_values = inf_data.posterior.stack(draws=("chain", "draw"))
+            posterior_values = posterior_values.lambdas.values.transpose()
+
+        return NapsuMQResult(mnjax, posterior_values, category_mapping)
+
+
+class NapsuMQResult(InferenceResult):
```

Comment:
There should be an easy way for the use to retrieve the queries that were used to get the result. You can get them from `MarkovNetworkJax`, but then the user will have to deal with `MarginalQuery` objects, which are not used to specify the queries.